src/Engines/BaseEngine.cs

```diff
@@ -139,6 +141,8 @@ public ExecutedEventArgs(ISymbol symbol, ExecutionResult result, TimeSpan durati
         private Dictionary<string, Stack<IResultEncoder>> serializers = new Dictionary<string, Stack<IResultEncoder>>();
         private List<ISymbolResolver> resolvers = new List<ISymbolResolver>();
 
+        private readonly Stack<Lazy<Task<ExecutionResult?>>> executionQueue = new Stack<Lazy<Task<ExecutionResult?>>>();
```

Comment:
I don't quite understand why we have an stack (last in, first out) that we are calling and using as a queue (first in, first out).
I guess the idea was to stack the task results so we can wait for them in the reverse order, but I think this was getting too complicated...

I wrote an example of how I would do it here:
https://github.com/vxfield/AsyncMessageHandler

You could continue to have the handler to be just an Action, or you could make the handler to be a subclass of an abstract MessageHandler class and the protected execute method.
Having it as a separate class might be better to encapsulate the handlers to their own "component" and their could set their individual message handling mode (see my example).

Reply:
Thanks for your feedback! I've added a separate class to represent async handlers that need strict ordering of their messages, following your example. I think that cleaned up the code quite a bit.  Please let me know if you think there's anything else I should add. Thanks!